From 786cc7c225335ba01fc50743e1612d2faed47938 Mon Sep 17 00:00:00 2001
From: Johannes Thoma <johannes@johannesthoma.com>
Date: Mon, 28 May 2018 18:31:51 +0200
Subject: [PATCH 2/2] Create/remove windows device on becoming
 primary/secondary.

That way we avoid caching modify/modify conflicts with the
Windows file cache.
---
 drbd/drbd_nl.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/drbd/drbd_nl.c b/drbd/drbd_nl.c
index f772f76..e2e9a8f 100644
--- a/drbd/drbd_nl.c
+++ b/drbd/drbd_nl.c
@@ -1008,10 +1008,13 @@ drbd_set_role(struct drbd_resource *resource, enum drbd_role role, bool force, s
 	enum chg_state_flags flags = CS_ALREADY_SERIALIZED | CS_DONT_RETRY | CS_WAIT_COMPLETE;
 	enum drbd_role old_role = resource->role[NOW];
 
-	if (old_role == R_PRIMARY && role == R_SECONDARY)
-		idr_for_each_entry(&resource->devices, device, vnr)
+	if (old_role == R_PRIMARY && role == R_SECONDARY) {
+		idr_for_each_entry(&resource->devices, device, vnr) {
 			if (windrbd_umount(device->this_bdev) != 0)
-				drbd_warn(device, "Couldn't umount volume %d\n", vnr);
+				drbd_warn(device, "couldn't umount volume %d\n", vnr);
+			windrbd_remove_windows_device(device->this_bdev);
+		}
+	}
 
 retry:
 	down(&resource->state_sem);
@@ -1239,11 +1242,16 @@ retry:
 			drbd_kobject_uevent(device);
 	}
 
-	if (old_role == R_SECONDARY && role == R_PRIMARY)
-		idr_for_each_entry(&resource->devices, device, vnr)
+	if (old_role == R_SECONDARY && role == R_PRIMARY) {
+		idr_for_each_entry(&resource->devices, device, vnr) {
+			if (windrbd_create_windows_device(device->this_bdev) != 0)
+				drbd_warn(device, "Couldn't create windows device for volume %d\n", vnr);
 			if (windrbd_mount(device->this_bdev) != 0)
 				drbd_warn(device, "Couldn't mount volume %d\n", vnr);
 
+		}
+	}
+
 out:
 	up(&resource->state_sem);
 	if (err_str) {
-- 
2.7.4

